src/FWlib/peripheral_keyscan.c

```diff
@@ -0,0 +1,254 @@
+    return;
+}
+
+void KEYSCAN_SetReleaseTime(uint32_t time)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 11;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_scanner_ctrl1, offset, bits_width, time);
+    return;
+}
+
+void KEYSCAN_SetScanInterval(uint32_t scan_itv)
+{
+    uint8_t offset = 11;
+    uint8_t bits_width = 16;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_scanner_ctrl1, offset, bits_width, scan_itv);
+    return;
+}
+
+void KEYSCAN_SetOutRowMask(uint32_t row_mask)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 20;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_row_mask_ctrl, offset, bits_width, row_mask);
+    return;
+}
+
+void KEYSCAN_SetInColMask(uint32_t col_mask)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 20;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_col_mask_ctrl, offset, bits_width, col_mask);
+    return;
+}
+
+void KEYSCAN_SetIntTrigEn(uint8_t enable)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 1;
+    uint32_t data = (enable == 0) ? 0: 1;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_int_en, offset, bits_width, data);
+    return;
+}
+
+void KEYSCAN_SetFifoClrReg(uint8_t set)
+{
+    uint8_t offset = 4;
+    uint8_t bits_width = 1;
+    uint32_t data = (set == 0) ? 0: 1;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_int_en, offset, bits_width, data);
+    return;
+}
+
+uint8_t KEYSCAN_GetIntStateFifoEmptyRaw(void)
+{
+    uint8_t offset = 6;
+    uint8_t bits_width = 1;
+    uint8_t ret;
+
+    ret = KEYSCAN_reg_read_bits(&APB_KEYSCAN->key_int, offset, bits_width);
+    return ret;
+}
+
+uint8_t KEYSCAN_GetKeyScannerBusy(void)
+{
+    uint8_t offset = 6;
+    uint8_t bits_width = 1;
+    uint8_t ret;
+
+    ret = KEYSCAN_reg_read_bits(&APB_KEYSCAN->fifo_status, offset, bits_width);
+    return ret;
+}
+
+uint16_t KEYSCAN_GetKeyData(void)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 11;
+    uint16_t ret;
+
+    ret = KEYSCAN_reg_read_bits(&APB_KEYSCAN->key_data, offset, bits_width);
+    return ret;
+}
+
+void KEYSCAN_SetFifoNumTrigInt(uint32_t trig_num)
+{
+    uint8_t offset = 0;
+    uint8_t bits_width = 5;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_trig, offset, bits_width, trig_num);
+    return;
+}
+
+void KEYSCAN_SetDmaNumTrigInt(uint32_t trig_num)
+{
+    uint8_t offset = 5;
+    uint8_t bits_width = 5;
+
+    KEYSCAN_reg_write_bits(&APB_KEYSCAN->key_trig, offset, bits_width, trig_num);
+    return;
+}
+
+static void KEYSCAN_InitKeyScanToIdx(KEYSCAN_SetStateStruct* keyscan_set)
+{
+    int i;
+
+    for (i = 0; i < keyscan_set->col_num; i++) {
+        keyscan_set->ctx->col_to_idx[keyscan_set->col[i].in_col] =  i;
+    }
+
+    for (i = 0; i < keyscan_set->row_num; i++) {
+        keyscan_set->ctx->row_to_idx[keyscan_set->row[i].out_row] =  i;
+    }
+
+    return;
+}
+
+uint8_t KEYSCAN_KeyDataToRowColIdx(KEYSCAN_SetStateStruct* keyscan_set, uint32_t key_data, uint8_t *row, uint8_t *col)
+{
+    if (key_data == 0x00000400) {
+        return 0;
+    }
```

Comment:
(key_data >= 0x00000400) ?

Reply:
At the end of every loop, 0x400 will be pulled into the FIFO